### PR TITLE
feat(proposer): check tko balance and fee before proposing

### LINKED ISSUE #203: feat(proposer): check current proposer's Taiko token balance && block fee before each proposing operation

### PR DESCRIPTION
**Pull request by cyberhorsey**
closes #203 
sorry roger, i thought this was assigned me to but had the wrong issue open :D.

tests are not failing due to my code changes, perhaps genesis issue?

### CODE REVIEW

File: proposer/proposer.go [hunk 1]
```diff
@@ -166,6 +168,12 @@ func (p *Proposer) ProposeOp(ctx context.Context) error {
 		return fmt.Errorf("failed to wait until L2 execution engine synced: %w", err)
 	}
 
+	log.Info("Comparing proposer TKO balance to block fee")
+
+	if err := p.checkTaikoTokenBalance(); err != nil {
```
Comment (davidtaikocha) [inline proposer/proposer.go:173]: hmmm check the balance when initing or before each proposing operation, not sure which approach is better?

**Comment by davidtaikocha** [comment]
> tests are not failing due to my code changes, perhaps genesis issue?

Its a bindings issue, updated.

File: proposer/proposer.go (hunk 1 above)
Comment (cyberhorsey) [reply proposer/proposer.go:173]: i moved it to ProposeTxList, is that better?

Comment (davidtaikocha) [reply proposer/proposer.go]: will move it to `ProposeOp` better? if the proposer doesn't have enough balance, we won't waste a RPC call to fetch the mempool?

Comment (cyberhorsey) [reply proposer/proposer.go:168]: OK!

**Review by davidtaikocha — APPROVED** [review]

**Pull request merged**